src/Workspaces/Core/Portable/Classification/SyntaxClassification/SyntacticChangeRangeComputer.cs

```diff
@@ -0,0 +1,220 @@
+            leftNewStack.Object.Push(newRoot);
+            rightOldStack.Object.Push(oldRoot);
+            rightNewStack.Object.Push(newRoot);
+
+            // We will be comparing the trees for two documents like so:
+            //
+            //       --------------------------------------------
+            //  old: |                                          |
+            //       --------------------------------------------
+            //
+            //       ---------------------------------------------------
+            //  new: |                                                 |
+            //       ---------------------------------------------------
+            //
+            // (Note that `new` could be smaller or the same length as `old`, it makes no difference).
+            //
+            // The algorithm will sweep in from both sides, as long as the nodes and tokens it's touching on each side
+            // are 'identical' (i.e. are the exact same green node, and were thus reused over an incremental parse.).
+            // This will leave us with:
+            //
+            //       --------------------------------------------
+            //  old: |  CLW  |                      |    CRW    |
+            //       --------------------------------------------
+            //       |       |                          \           \
+            //       ---------------------------------------------------
+            //  new: |  CLW  |                             |    CRW    |
+            //       ---------------------------------------------------
+            //
+            // Where CLW and CRW refer to the common-left-width and common-right-width respectively. The part in between
+            // this s the change range:
+            //
+            //       --------------------------------------------
+            //  old: |       |**********************|           |
+            //       --------------------------------------------
+            //               |**************************\
+            //       ---------------------------------------------------
+            //  new: |       |*****************************|           |
+            //       ---------------------------------------------------
+            //
+            // The Span changed will go from `[CLW, Old_Width - CRW)`, and the NewLength will be `New_Width - CLW - CRW`
+
+            var commonLeftWidth = ComputeCommonLeftWidth(oldRoot, newRoot, leftOldStack.Object, leftNewStack.Object, cancellationToken);
+            var commonRightWidth = ComputeCommonRightWidth(oldRoot, newRoot, rightOldStack.Object, rightNewStack.Object, cancellationToken);
+
+            if (commonLeftWidth == null || commonRightWidth == null)
+            {
+                // They better have both been the same, otherwise our walk was somehow non-symmetric.
+                Contract.ThrowIfFalse(commonLeftWidth == commonRightWidth);
```

Comment:
💡 Since this path is calling `ComputeCommonRightWidth` unnecessarily, consider adding a comment explaining that this is not the hot path: for any text edit aside from overtyping a character with _exactly the same character_, there will be at least one edit somewhere. The value provided by runtime verification of algorithm consistency outweighs the negligible performance hit.

Reply:
not exactly sure i understood this comment.  this contract callwas mostly about me ensuring that certain invariants i believe to be true about trees are in fact true.